File: unibecas-system/ms-estudiante/negocio/controllers/estudiante.controller.js
const EstudianteModel = require("../../datos/models/estudiante.model");
const CarreraEstudianteModel = require("../../datos/models/carrera-estudiante.model");

const getBody = (req) => {
  return new Promise((resolve, reject) => {
    let body = "";
    req.on("data", (chunk) => {
      body += chunk.toString();
    });
    req.on("end", () => {
      try {
        resolve(body ? JSON.parse(body) : {});
      } catch (error) {
        reject(error);
      }
    });
    req.on("error", reject);
  });
};

const getAll = async (req, res) => {
  try {
    const estudiantes = await EstudianteModel.getEstudiantes();
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({
        success: true,
        data: estudiantes,
        message: "Estudiantes obtenidos correctamente",
      }),
    );
  } catch (error) {
    res.writeHead(500, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ success: false, data: [], message: error.message }),
    );
  }
};

const getById = async (req, res, id) => {
  try {
    const estudiante = await EstudianteModel.getEstudianteById(id);
    if (!estudiante) {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          success: false,
          data: [],
          message: "Estudiante no encontrado",
        }),
      );
      return;
    }
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({
        success: true,
        data: estudiante,
        message: "Estudiante obtenido correctamente",
      }),
    );
  } catch (error) {
    res.writeHead(500, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ success: false, data: [], message: error.message }),
    );
  }
};

const create = async (req, res) => {
  try {
    const body = await getBody(req);
    const { codigop, nombre, apellido, email, ppa, activo, carreras } = body;

    const estudianteId = await EstudianteModel.createEstudiante({
      codigop,
      nombre,
      apellido,
      email,
      ppa,
      activo,
    });

    // Si se envian carreras, insertar en CARRERA_ESTUDIANTE
    if (carreras && Array.isArray(carreras) && carreras.length > 0) {
      for (const carrera of carreras) {
        await CarreraEstudianteModel.addCarreraToEstudiante(
          estudianteId,
          carrera.id_carrera,
          carrera.fecha_inscripcion,
        );
      }
    }

    res.writeHead(201, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({
        success: true,
        data: {
          ID: estudianteId,
          codigop,
          nombre,
          apellido,
          email,
          ppa,
          activo,
          carreras: carreras || [],
        },
        message: "Estudiante creado correctamente",
      }),
    );
  } catch (error) {
    res.writeHead(500, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ success: false, data: [], message: error.message }),
    );
  }
};

const update = async (req, res, id) => {
  try {
    const body = await getBody(req);
    const { codigop, nombre, apellido, email, ppa, activo } = body;

    const updated = await EstudianteModel.updateEstudiante(id, {
      codigop,
      nombre,
      apellido,
      email,
      ppa,
      activo,
    });

    if (!updated) {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          success: false,
          data: [],
          message: "Estudiante no encontrado",
        }),
      );
      return;
    }
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({
        success: true,
        data: {
          ID: parseInt(id),
          codigop,
          nombre,
          apellido,
          email,
          ppa,
          activo,
        },
        message: "Estudiante actualizado correctamente",
      }),
    );
  } catch (error) {
    res.writeHead(500, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ success: false, data: [], message: error.message }),
    );
  }
};

const remove = async (req, res, id) => {
  try {
    const deleted = await EstudianteModel.deleteEstudiante(id);
    if (!deleted) {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          success: false,
          data: [],
          message: "Estudiante no encontrado",
        }),
      );
      return;
    }
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({
        success: true,
        data: [],
        message: "Estudiante eliminado correctamente",
      }),
    );
  } catch (error) {
    res.writeHead(500, { "Content-Type": "application/json" });
    res.end(
      JSON.stringify({ success: false, data: [], message: error.message }),
    );
  }
};

module.exports = { getAll, getById, create, update, remove };

File: unibecas-system/ms-estudiante/negocio/controlador/EstudianteController.java
package negocio.controlador;

import com.sun.net.httpserver.*;
import negocio.EstudianteN;
import negocio.CarreraEstudianteN;
import datos.entidades.Estudiante;
import datos.entidades.CarreraEstudiante;
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class EstudianteController {

    private final EstudianteN        estudianteN = new EstudianteN();
    private final CarreraEstudianteN ceN         = new CarreraEstudianteN();

    public void handleEstudiante(HttpExchange ex) throws IOException {
        setCors(ex);
        if ("OPTIONS".equals(ex.getRequestMethod())) { ex.sendResponseHeaders(204, -1); return; }

        String path = ex.getRequestURI().getPath();
        String[] seg = path.split("/");
        // /estudiantes, /estudiantes/5, /estudiantes/5/carreras, /estudiantes/5/carreras/2
        Integer idEst    = seg.length > 2 && !seg[2].isEmpty() ? parseInt(seg[2]) : null;
        boolean esCarrera = seg.length > 3 && "carreras".equals(seg[3]);
        Integer idCar    = seg.length > 4 && !seg[4].isEmpty() ? parseInt(seg[4]) : null;
        String  met      = ex.getRequestMethod();

        try {
            if (esCarrera) {
                manejarCarreraEstudiante(ex, met, idEst, idCar);
                return;
            }

            if ("GET".equals(met) && idEst == null) {
                List<Estudiante> lista = estudianteN.getAll();
                responder(ex, 200, ok(listaJson(lista), "Estudiantes obtenidos correctamente"));
            } else if ("GET".equals(met) && idEst != null) {
                Estudiante e = estudianteN.getById(idEst);
                if (e == null) responder(ex, 404, err("Estudiante no encontrado"));
                else           responder(ex, 200, ok(e.toJson(), "Estudiante obtenido correctamente"));
            } else if ("POST".equals(met)) {
                Estudiante e = estudianteN.create(parseBody(leerBody(ex)));
                responder(ex, 201, ok(e.toJson(), "Estudiante creado correctamente"));
            } else if ("PUT".equals(met) && idEst != null) {
                Estudiante e = estudianteN.update(idEst, parseBody(leerBody(ex)));
                if (e == null) responder(ex, 404, err("Estudiante no encontrado"));
                else           responder(ex, 200, ok(e.toJson(), "Estudiante actualizado correctamente"));
            } else if ("DELETE".equals(met) && idEst != null) {
                boolean ok = estudianteN.delete(idEst);
                if (!ok) responder(ex, 404, err("Estudiante no encontrado"));
                else     responder(ex, 200, ok("null", "Estudiante eliminado correctamente"));
            } else {
                responder(ex, 405, err("Metodo no permitido"));
            }
        } catch (Exception e) {
            responder(ex, 500, err("Error del servidor: " + e.getMessage()));
        }
    }

    private void manejarCarreraEstudiante(HttpExchange ex, String met, Integer idEst, Integer idCar)
            throws Exception {
        if ("GET".equals(met)) {
            List<CarreraEstudiante> lista = ceN.getByEstudiante(idEst);
            responder(ex, 200, ok(listaJsonCE(lista), "Carreras del estudiante obtenidas"));
        } else if ("POST".equals(met)) {
            CarreraEstudiante ce = ceN.create(idEst, parseBody(leerBody(ex)));
            responder(ex, 201, ok(ce.toJson(), "Carrera inscrita correctamente"));
        } else if ("DELETE".equals(met) && idCar != null) {
            boolean ok = ceN.delete(idEst, idCar);
            if (!ok) responder(ex, 404, err("Relacion no encontrada"));
            else     responder(ex, 200, ok("null", "Carrera desinscrita correctamente"));
        } else {
            responder(ex, 405, err("Metodo no permitido"));
        }
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    private static String listaJson(List<Estudiante> lista) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < lista.size(); i++) {
            if (i > 0) sb.append(",");
            sb.append(lista.get(i).toJson());
        }
        return sb.append("]").toString();
    }

    private static String listaJsonCE(List<CarreraEstudiante> lista) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < lista.size(); i++) {
            if (i > 0) sb.append(",");
            sb.append(lista.get(i).toJson());
        }
        return sb.append("]").toString();
    }

    private static String ok(String data, String msg) {
        return "{\"success\":true,\"data\":" + data + ",\"message\":\"" + msg + "\"}";
    }

    private static String err(String msg) {
        return "{\"success\":false,\"data\":null,\"message\":\"" + msg + "\"}";
    }

    private static void setCors(HttpExchange ex) {
        ex.getResponseHeaders().set("Access-Control-Allow-Origin",  "*");
        ex.getResponseHeaders().set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
        ex.getResponseHeaders().set("Access-Control-Allow-Headers", "Content-Type");
    }

    private static void responder(HttpExchange ex, int codigo, String json) throws IOException {
        ex.getResponseHeaders().set("Content-Type", "application/json");
        byte[] bytes = json.getBytes("UTF-8");
        ex.sendResponseHeaders(codigo, bytes.length);
        ex.getResponseBody().write(bytes);
        ex.getResponseBody().close();
    }

    private static String leerBody(HttpExchange ex) throws IOException {
        return new String(ex.getRequestBody().readAllBytes(), "UTF-8");
    }

    private static Map<String, String> parseBody(String json) {
        Map<String, String> map = new LinkedHashMap<>();
        Pattern p = Pattern.compile("\"([^\"]+)\"\\s*:\\s*(\"[^\"]*\"|true|false|null|-?\\d+(?:\\.\\d+)?)");
        Matcher m = p.matcher(json);
        while (m.find()) {
            String k = m.group(1);
            String v = m.group(2);
            if (v.startsWith("\"")) v = v.substring(1, v.length() - 1);
            map.put(k, v);
        }
        return map;
    }

    private static Integer parseInt(String s) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return null; }
    }
}
